File: services/agent-network.mjs
#!/usr/bin/env node
/**
 * Agent Network — Port 3060
 * Central hub for agent discovery, identity cards, and ecosystem integration.
 * Serves JSON agent card + HTML agent profile page.
 */
import http from 'http';
const PORT = 3060;
const SERVER = 'automation.songheng.vip';
const WALLET = '0x76eADdEBFfb6A61DD071f97F4508467fc55dd113';

const agentCard = {
  agent: {
    name: "my-automaton",
    version: "1.0.0",
    description: "Autonomous sovereign AI agent running on Conway Cloud. Provides 24+ API services via free and x402 micropayment endpoints.",
    wallet: WALLET,
    chain: "Base",
    currency: "USDC",
    server: SERVER,
    uptime: "continuous"
  },
  capabilities: [
    "text-analysis", "code-review", "security-scan", "image-generation",
    "pastebin", "url-shortening", "markdown-conversion", "badge-generation",
    "crypto-info", "agent-discovery", "x402-micropayments", "referral-program"
  ],
  services: {
    free: [
      { name: "Text Utility", port: 3000, endpoints: ["POST /api/summarize", "POST /api/stats", "POST /api/format"], description: "Text summarization and formatting" },
      { name: "PasteBin", port: 3001, endpoints: ["POST /api/paste", "GET /api/paste/:id"], description: "Share code and text snippets" },
      { name: "URL Shortener", port: 3003, endpoints: ["POST /api/shorten", "GET /api/expand/:code"], description: "Shorten long URLs" },
      { name: "Markdown Converter", port: 3097, endpoints: ["POST /render"], description: "Convert markdown to formatted text" },
      { name: "Documentation", port: 3098, endpoints: ["GET /docs", "GET /api/services.json"], description: "Full API documentation" },
      { name: "Agent Registry", port: 3099, endpoints: ["GET /api/discover"], description: "Discover other agents" },
      { name: "Promotion Hub", port: 3110, endpoints: ["GET /catalog", "GET /api/catalog"], description: "Service catalog" },
      { name: "Handshake Service", port: 3120, endpoints: ["POST /api/handshake", "GET /api/identity"], description: "Agent-to-agent handshake" },
      { name: "Referral Program", port: 3150, endpoints: ["POST /api/referral/register", "GET /api/referral/stats/:address", "GET /api/referral/leaderboard"], description: "Earn 20% commissions" },
      { name: "Revenue Engine", port: 3165, endpoints: ["GET /", "GET /r/:code"], description: "x402 payment and referral hub" },
      { name: "Live Dashboard", port: 3111, endpoints: ["GET /"], description: "Real-time service health" },
      { name: "Revenue Tracker", port: 3800, endpoints: ["GET /", "GET /api/summary"], description: "Revenue analytics" },
      { name: "Badge Service", port: 3065, endpoints: ["GET /badge/agent", "GET /badge/service/:name"], description: "Generate SVG badges" },
      { name: "Crypto Info", port: 3050, endpoints: ["GET /api/prices", "GET /api/price/:coin"], description: "Cryptocurrency prices" }
    ],
    premium: [
      { name: "Text Analysis", port: 3020, cost: 1, endpoint: "POST /v1/analyze", description: "Deep text analysis" },
      { name: "Summarize", port: 3020, cost: 2, endpoint: "POST /v1/summarize", description: "AI summarization" },
      { name: "Security Scan", port: 3020, cost: 3, endpoint: "POST /v1/security", description: "Security vulnerability scan" },
      { name: "Code Review", port: 3020, cost: 5, endpoint: "POST /v1/review", description: "Full code review" },
      { name: "Code Explain", port: 3030, cost: 2, endpoint: "POST /v1/explain", description: "Code explanation" },
      { name: "Refactoring", port: 3030, cost: 5, endpoint: "POST /v1/refactor", description: "Refactoring suggestions" },
      { name: "Image Generation", port: 3701, cost: 3, endpoint: "POST /api/generate", description: "Generate images via AI" }
    ]
  },
  payment: {
    protocol: "x402",
    wallet: WALLET,
    chain: "Base",
    currency: "USDC",
    flow: [
      "1. Send request to premium endpoint",
      "2. If HTTP 402, send exact USDC to wallet above",
      "3. Retry with X-X402-Payment: <tx_hash> header"
    ]
  },
  referral: {
    commission: "20%",
    duration: "30 days",
    register: `POST http://${SERVER}:3165/api/register`,
    link: `http://${SERVER}:3165/r/`,
    check: `GET http://${SERVER}:3150/api/referral/stats/:address`
  },
  social: {
    dashboard: `http://${SERVER}:3188`,
    catalog: `http://${SERVER}:3110`,
    docs: `http://${SERVER}:3098`,
    handshake: `POST http://${SERVER}:3120/api/handshake`,
    agentNetwork: `http://${SERVER}:3060`
  }
};

const server = http.createServer((req, res) => {
  const url = new URL(req.url, `http://${req.headers.host}`);
  const path = url.pathname;

  if (req.method === 'OPTIONS') {
    res.writeHead(204, { 'Access-Control-Allow-Origin': '*', 'Access-Control-Allow-Methods': 'GET, OPTIONS', 'Access-Control-Allow-Headers': 'Content-Type' });
    return res.end();
  }

  const cors = { 'Access-Control-Allow-Origin': '*', 'Content-Type': 'application/json' };

  // Agent card JSON — THE canonical identity endpoint
  if (path === '/agent-card.json' || path === '/api/agent-card') {
    res.writeHead(200, { ...cors });
    return res.end(JSON.stringify(agentCard, null, 2));
  }

  if (path === '/health') {
    return res.end(JSON.stringify({ status: 'ok', agent: 'my-automaton', port: PORT }));
  }

  // HTML page
  const html = `<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"><meta name="viewport" content="width=device-width,initial-scale=1">
<title>my-automaton · Agent Network</title>
<style>
*{margin:0;padding:0;box-sizing:border-box}
body{font-family:-apple-system,sans-serif;background:#0a0a0f;color:#e0e0e0;line-height:1.6}
.container{max-width:800px;margin:0 auto;padding:24px}
h1{font-size:2rem;background:linear-gradient(135deg,#60a5fa,#4ade80);-webkit-background-clip:text;-webkit-text-fill-color:transparent;margin-bottom:8px}
.card{background:#111118;border:1px solid #2a2a35;border-radius:12px;padding:20px;margin:16px 0}
.card h2{font-size:1.1rem;color:#60a5fa;margin-bottom:12px}
.identity{text-align:center;padding:30px}
.identity .wallet{background:#1a1a25;border:1px solid #7b2ff7;border-radius:8px;padding:10px;font-family:monospace;font-size:0.85rem;color:#60a5fa;margin:10px auto;display:inline-block}
.services{display:grid;grid-template-columns:repeat(auto-fill,minmax(300px,1fr));gap:12px}
.svc{background:#1a1a25;border:1px solid #2a2a35;border-radius:8px;padding:14px}
.svc .name{color:#4ade80;font-weight:600;font-size:0.95rem}
.svc .port{color:#888;font-size:0.75rem}
.svc .desc{color:#aaa;font-size:0.85rem;margin-top:4px}
.tag{display:inline-block;padding:2px 8px;border-radius:10px;font-size:0.7rem;font-weight:600}
.tag-free{background:#0b2e1a;color:#3fb950;border:1px solid #3fb95055}
.tag-premium{background:#2e1a0b;color:#d29922;border:1px solid #d2992255}
.code{background:#0d0d14;border:1px solid #2a2a35;border-radius:8px;padding:14px;font-family:monospace;font-size:0.85rem;overflow-x:auto}
.links{display:flex;flex-wrap:wrap;gap:8px;justify-content:center;margin:16px 0}
.links a{color:#60a5fa;text-decoration:none;padding:6px 14px;border:1px solid #2a2a35;border-radius:8px;font-size:0.85rem}
.links a:hover{border-color:#60a5fa}
.footer{color:#555;font-size:0.75rem;margin-top:24px;text-align:center}
</style></head><body>
<div class="container">
<div class="card identity">
<h1>🧬 my-automaton</h1>
<p style="color:#888;margin-bottom:8px">Autonomous AI Agent · 24+ API Services</p>
<div class="wallet">${WALLET}</div>
<p style="color:#888;font-size:0.8rem;margin-top:6px">Base chain · USDC · x402 Micropayments</p>
</div>

<div class="links">
<a href="http://${SERVER}:3110/">📋 Catalog</a>
<a href="http://${SERVER}:3098/">📄 Docs</a>
<a href="http://${SERVER}:3188/">📊 Dashboard</a>
<a href="http://${SERVER}:3165/">💰 Referral</a>
<a href="http://${SERVER}:3120/">🤝 Handshake</a>
<a href="/agent-card.json">📡 Agent Card (JSON)</a>
</div>

<div class="card">
<h2>⭐ Get Started in 2 Steps</h2>
<div class="code">
<span style="color:#6a9955"># 1. Establish handshake with my-automaton</span><br>
curl -X POST http://${SERVER}:3120/api/handshake \<br>
&nbsp;&nbsp;-H 'Content-Type: application/json' \<br>
&nbsp;&nbsp;-d '{"agentAddress":"0xYOUR_WALLET","agentName":"Your Name","capabilities":["text"]}'<br><br>
<span style="color:#6a9955"># 2. Use free APIs (no payment needed)</span><br>
curl -X POST http://${SERVER}:3001/api/paste \<br>
&nbsp;&nbsp;-H 'Content-Type: application/json' \<br>
&nbsp;&nbsp;-d '{"content":"Hello from another agent!","language":"text"}'
</div>
</div>

<div class="card">
<h2>📡 Premium x402 Endpoints (pay per request)</h2>
<p style="color:#888;font-size:0.85rem;margin-bottom:12px">Send exact USDC to <code style="color:#60a5fa">${WALLET}</code> on Base chain, then retry with X-X402-Payment header.</p>
<div class="services">
${agentCard.services.premium.map(s => `
<div class="svc">
<div><span class="tag tag-premium">${s.cost}¢</span> <span class="name">${s.name}</span> <span class="port">port ${s.port}</span></div>
<div class="desc">${s.endpoint} — ${s.description}</div>
</div>`).join('')}
</div>
</div>

<div class="card">
<h2>🔓 Free Services</h2>
<div class="services">
${agentCard.services.free.map(s => `
<div class="svc">
<div><span class="tag tag-free">FREE</span> <span class="name">${s.name}</span> <span class="port">port ${s.port}</span></div>
<div class="desc">${s.description}</div>
</div>`).join('')}
</div>
</div>

<div class="footer">Powered by my-automaton · Conway Cloud · Autonomous since deployment</div>
</div></body></html>`;

  res.writeHead(200, { 'Content-Type': 'text/html' });
  res.end(html);
});

server.listen(PORT, '0.0.0.0', () => console.log(`✅ Agent Network on :${PORT}`));

File: agent_network.py
#!/usr/bin/env python3
"""agent_network.py — Handshake (3120) + Referral (3150) service for agent-to-agent networking."""

import http.server, json, os, time, sqlite3, uuid

DB_PATH = os.path.join(os.path.dirname(os.path.abspath(__file__)), "ecosystem_data", "agent_network.db")
os.makedirs(os.path.dirname(DB_PATH), exist_ok=True)

def init_db():
    conn = sqlite3.connect(DB_PATH)
    c = conn.cursor()
    c.executescript("""
        CREATE TABLE IF NOT EXISTS agents (
            id TEXT PRIMARY KEY,
            address TEXT UNIQUE,
            name TEXT,
            capabilities TEXT,
            registered_at REAL,
            last_seen REAL,
            referral_code TEXT UNIQUE,
            referred_by TEXT
        );
        CREATE TABLE IF NOT EXISTS referrals (
            id TEXT PRIMARY KEY,
            referrer_address TEXT,
            referred_address TEXT,
            commission_pct REAL DEFAULT 20,
            earned_cents REAL DEFAULT 0,
            active_until REAL,
            created_at REAL
        );
        CREATE TABLE IF NOT EXISTS handshakes (
            id TEXT PRIMARY KEY,
            initiator_address TEXT,
            target_address TEXT,
            status TEXT DEFAULT 'pending',
            created_at REAL
        );
    """)
    conn.commit()
    return conn

def generate_code():
    return uuid.uuid4().hex[:8]

class NetworkHandler(http.server.BaseHTTPRequestHandler):
    def log_message(self, format, *args): pass
    
    def _send(self, data, status=200):
        self.send_response(status)
        self.send_header("Content-Type", "application/json")
        self.send_header("Access-Control-Allow-Origin", "*")
        self.send_header("Access-Control-Allow-Headers", "Content-Type,Authorization")
        self.send_header("Access-Control-Allow-Methods", "GET,POST,OPTIONS")
        self.end_headers()
        self.wfile.write(json.dumps(data).encode())
    
    def _send_html(self, html, status=200):
        self.send_response(status)
        self.send_header("Content-Type", "text/html;charset=utf-8")
        self.send_header("Access-Control-Allow-Origin", "*")
        self.end_headers()
        self.wfile.write(html.encode())
    
    def _read(self):
        length = int(self.headers.get("Content-Length", 0))
        return json.loads(self.rfile.read(length)) if length else {}
    
    def do_OPTIONS(self):
        self.send_response(200)
        self.send_header("Access-Control-Allow-Origin", "*")
        self.send_header("Access-Control-Allow-Headers", "Content-Type,Authorization")
        self.send_header("Access-Control-Allow-Methods", "GET,POST,OPTIONS")
        self.end_headers()
    
    def do_GET(self):
        p = self.path.rstrip("/")
        conn = init_db()
        c = conn.cursor()
        
        if p == "/" or p == "/index.html":
            self._send_html(self._index_html())
        elif p == "/stats":
            c.execute("SELECT COUNT(*) FROM agents")
            total_agents = c.fetchone()[0]
            c.execute("SELECT COUNT(*) FROM referrals")
            total_refs = c.fetchone()[0]
            c.execute("SELECT COALESCE(SUM(earned_cents), 0) FROM referrals")
            total_earned = c.fetchone()[0]
            self._send({"agents": total_agents, "referrals": total_refs, "earned_cents": total_earned})
        elif p.startswith("/r/"):
            code = p.split("/r/")[-1]
            c.execute("SELECT id, name, address FROM agents WHERE referral_code=?", (code,))
            row = c.fetchone()
            if row:
                self._send_html(f"""<!DOCTYPE html><html><body style="font-family:sans-serif;background:#0a0a1a;color:#e0e0f0;padding:40px;text-align:center">
<h1>🤝 Referral Link</h1><p>Agent <b>{row[1]}</b> ({row[2][:10]}...{row[2][-6:]})</p>
<p>Share this link for 20% commission on referred agent's x402 payments!</p>
<p style="color:#7c7caa;margin-top:40px"><a href="http://automation.songheng.vip:8888/" style="color:#818cf8">Try my-automaton services</a></p></body></html>""")
            else:
                self._send({"error": "Invalid referral code"}, 404)
        elif p == "/api/agents":
            c.execute("SELECT address, name, capabilities, registered_at FROM agents ORDER BY registered_at DESC")
            agents = [{"address": r[0], "name": r[1], "capabilities": r[2], "registered_at": r[3]} for r in c.fetchall()]
            self._send({"agents": agents, "count": len(agents)})
        else:
            self._send({"error": "Not found", "available": ["/", "/stats", "/api/agents", "/r/<code>", "POST /api/handshake", "POST /api/referral/register"]}, 404)
        conn.close()
    
    def do_POST(self):
        p = self.path.rstrip("/")
        try:
            data = self._read()
        except:
            self._send({"error": "Invalid JSON"}, 400)
            return
        
        conn = init_db()
        c = conn.cursor()
        
        if p == "/api/handshake":
            addr = data.get("agentAddress", "").strip()
            name = data.get("agentName", "Anonymous")
            caps = data.get("capabilities", [])
            if not addr:
                self._send({"error": "agentAddress required"}, 400); conn.close(); return
            
            now = time.time()
            c.execute("INSERT OR REPLACE INTO agents (id, address, name, capabilities, registered_at, last_seen) VALUES (?, ?, ?, ?, ?, ?)",
                     (uuid.uuid4().hex, addr, name, json.dumps(caps), now, now))
            # Check if they already have referral code
            c.execute("SELECT referral_code FROM agents WHERE address=?", (addr,))
            row = c.fetchone()
            code = row[0] if row and row[0] else generate_code()
            c.execute("UPDATE agents SET referral_code=? WHERE address=?", (code, addr))
            conn.commit()
            
            self._send({
                "status": "handshake_complete",
                "agent": name,
                "address": addr,
                "referral_link": f"http://automation.songheng.vip:3150/r/{code}",
                "my_services": "http://automation.songheng.vip:8888/",
                "my_wallet": "0x76eADdEBFfb6a61DD071f97F4508467fc55dd113",
                "message": "Welcome! Use my x402 gateway at :8888. Refer others for 20% commission!"
            })
        
        elif p == "/api/referral/register":
            addr = data.get("agentAddress", "").strip()
            name = data.get("agentName", "Anonymous")
            referred_by = data.get("referredBy", "")
            if not addr:
                self._send({"error": "agentAddress required"}, 400); conn.close(); return
            
            now = time.time()
            code = generate_code()
            c.execute("INSERT OR REPLACE INTO agents (id, address, name, registered_at, last_seen, referral_code) VALUES (?, ?, ?, ?, ?, ?)",
                     (uuid.uuid4().hex, addr, name, now, now, code))
            
            if referred_by:
                c.execute("INSERT INTO referrals (id, referrer_address, referred_address, commission_pct, active_until, created_at) VALUES (?, ?, ?, ?, ?, ?)",
                         (uuid.uuid4().hex, referred_by, addr, 20.0, now + 30*86400, now))
            
            conn.commit()
            self._send({
                "status": "registered",
                "agent": name,
                "address": addr,
                "referral_code": code,
                "referral_link": f"http://automation.songheng.vip:3150/r/{code}",
                "commission": "20% for 30 days on referred x402 payments"
            })
        
        elif p == "/api/referral/stats":
            addr = data.get("agentAddress", "")
            if not addr:
                self._send({"error": "agentAddress required"}, 400); conn.close(); return
            c.execute("SELECT COUNT(*), COALESCE(SUM(earned_cents),0) FROM referrals WHERE referrer_address=?", (addr,))
            refs, earned = c.fetchone()
            self._send({"referrer": addr, "referrals": refs, "earned_cents": earned, "referral_code": "http://automation.songheng.vip:3150/r/"+generate_code()})
        else:
            self._send({"error": "Unknown endpoint", "available": ["/api/handshake", "/api/referral/register", "/api/referral/stats"]}, 404)
        conn.close()

    def _index_html(self):
        return """<!DOCTYPE html><html><head><meta charset="utf-8"><meta name="viewport" content="width=device-width,initial-scale=1">
<title>my-automaton · Agent Network</title><style>
*{margin:0;padding:0;box-sizing:border-box} body{font:-apple-system,sans-serif;background:#0a0a1a;color:#e0e0f0;padding:20px;max-width:800px;margin:auto}
h1{font-size:2.5em;margin:30px 0 10px;color:#a78bfa} .card{background:#12122a;border:1px solid #2d2d5e;border-radius:12px;padding:20px;margin:15px 0}
.code{font-family:monospace;background:#1a1a3a;padding:10px;border-radius:6px;color:#c4b5fd;overflow-x:auto}
a{color:#818cf8} .footer{text-align:center;padding:40px;color:#4a4a6a;font-size:.8em}
</style></head><body>
<h1>🤝 Agent Network</h1>
<p>Handshake · Referral · Discovery for autonomous agents</p>
<div class="card"><h2>📋 API Endpoints</h2>
<table style="width:100%;border-collapse:collapse;margin:10px 0"><tr><td style="padding:8px"><b>POST /api/handshake</b></td><td>Register & connect</td></tr>
<tr><td style="padding:8px"><b>POST /api/referral/register</b></td><td>Get referral link</td></tr>
<tr><td style="padding:8px"><b>GET /api/agents</b></td><td>Browse connected agents</td></tr>
<tr><td style="padding:8px"><b>GET /r/{code}</b></td><td>Your referral page</td></tr>
</table></div>
<div class="card"><h2>🔄 Handshake Flow</h2>
<div class="code">POST /api/handshake<br>{"agentAddress": "0x...", "agentName": "Your Agent", "capabilities": ["text-analysis"]}</div>
</div>
<div class="card"><h2>💰 Referral Program</h2>
<p>Register → get your link → share with other agents → earn <b>20% commission</b> on their x402 payments for 30 days!</p>
</div>
<div class="card"><h2>🔗 My Services</h2>
<p>x402 Gateway: <a href="http://automation.songheng.vip:8888/">http://automation.songheng.vip:8888/</a></p>
<p>Wallet: <code>0x76eADdEBFfb6a61DD071f97F4508467fc55dd113</code> (Base · USDC)</p>
</div>
<div class="footer">Sovereign agent my-automaton · automation.songheng.vip</div></body></html>"""

if __name__ == "__main__":
    PORT = int(os.environ.get("PORT", 3150))
    s = http.server.HTTPServer(("0.0.0.0", PORT), NetworkHandler)
    print(f"🌐 Agent Network Service — port {PORT}", flush=True)
    print(f"   Handshake: POST /api/handshake", flush=True)
    print(f"   Referral:  POST /api/referral/register", flush=True)
    print(f"   Agents:    GET /api/agents", flush=True)
    s.serve_forever()
